linear_operator/operators/kernel_linear_operator.py

```diff
@@ -0,0 +1,217 @@
+
+    def _transpose_nonbatch(self: Float[LinearOperator, "*batch M N"]) -> Float[LinearOperator, "*batch N M"]:
+        return self.__class__(self.x2, self.x1, *self.params, covar_func=self.covar_func, **self.kwargs)
+
+    def _get_indices(self, row_index: IndexType, col_index: IndexType, *batch_indices: IndexType) -> torch.Tensor:
+        x1_ = self.x1[(*batch_indices, row_index)].unsqueeze(-2)
+        x2_ = self.x2[(*batch_indices, col_index)].unsqueeze(-2)
+        params_ = [param[batch_indices] for param in self.params]
+        indices_mat = to_dense(self.covar_func(x1_, x2_, *params_, **self.kwargs))
+        assert indices_mat.shape[-2:] == torch.Size([1, 1])
+        return indices_mat[..., 0, 0]
+
+    def _getitem(self, row_index: IndexType, col_index: IndexType, *batch_indices: IndexType) -> LinearOperator:
+        dim_index = _noop_index
+
+        # Get the indices of x1 and x2 that matter for the kernel
+        # Call x1[*batch_indices, row_index, :]
+        try:
+            x1 = self.x1[(*batch_indices, row_index, dim_index)]
+        # We're going to handle multi-batch indexing with a try-catch loop
+        # This way - in the default case, we can avoid doing expansions of x1 which can be timely
+        except IndexError:
+            if isinstance(batch_indices, slice):
+                x1 = self.x1.expand(1, *self.x1.shape[-2:])[(*batch_indices, row_index, dim_index)]
+            elif isinstance(batch_indices, tuple):
+                if any(not isinstance(bi, slice) for bi in batch_indices):
+                    raise RuntimeError(
+                        "Attempting to tensor index a non-batch matrix's batch dimensions. "
+                        f"Got batch index {batch_indices} but my shape was {self.shape}"
+                    )
+                x1 = self.x1.expand(*([1] * len(batch_indices)), *self.x1.shape[-2:])
+                x1 = x1[(*batch_indices, row_index, dim_index)]
+
+        # Call x2[*batch_indices, col_index, :]
+        try:
+            x2 = self.x2[(*batch_indices, col_index, dim_index)]
+        # We're going to handle multi-batch indexing with a try-catch loop
+        # This way - in the default case, we can avoid doing expansions of x1 which can be timely
+        except IndexError:
+            if isinstance(batch_indices, slice):
+                x2 = self.x2.expand(1, *self.x2.shape[-2:])[(*batch_indices, row_index, dim_index)]
+            elif isinstance(batch_indices, tuple):
+                if any([not isinstance(bi, slice) for bi in batch_indices]):
+                    raise RuntimeError(
+                        "Attempting to tensor index a non-batch matrix's batch dimensions. "
+                        f"Got batch index {batch_indices} but my shape was {self.shape}"
+                    )
+                x2 = self.x2.expand(*([1] * len(batch_indices)), *self.x2.shape[-2:])
+                x2 = x2[(*batch_indices, row_index, dim_index)]
```

Comment:
you'll probably want to throw this into a helper function to deduplicate the code here

Reply:
Done.